pkg/ratelimiter/rate_limiter.go

```diff
@@ -31,88 +37,106 @@ type Entry struct {
 	mutex sync.Mutex
 }
 
+// Limit controls token refilling for bucket entries
+type Limit struct {
+	// Maximum number of tokens that can be accumulated
+	MaxTokens uint16
+	// Number of tokens to refill per minute
+	RatePerMinute uint16
+}
+
+func (l Limit) Refill(entry *Entry, multiplier uint16) {
+	now := time.Now()
+	ratePerMinute := l.RatePerMinute * multiplier
+	maxTokens := l.MaxTokens * multiplier
+	entry.mutex.Lock()
+	defer entry.mutex.Unlock()
+	minutesSinceLastSeen := now.Sub(entry.lastSeen).Minutes()
+	if minutesSinceLastSeen > 0 {
+		// Only update the lastSeen if it has been >= 1 minute
+		// This allows for continuously sending nodes to still get credits
+		entry.lastSeen = now
+		// Convert to ints so that we can check if above MAX_UINT_16
+		additionalTokens := int(ratePerMinute) * int(minutesSinceLastSeen)
+		// Avoid overflows of UINT16 when new balance is above limit
+		if additionalTokens+int(entry.tokens) > MAX_UINT_16 {
+			additionalTokens = MAX_UINT_16 - int(entry.tokens)
+		}
+		entry.tokens = minUint16(entry.tokens+uint16(additionalTokens), maxTokens)
+	}
+}
+
 // TokenBucketRateLimiter implements the RateLimiter interface
 type TokenBucketRateLimiter struct {
-	log     *zap.Logger
-	wallets map[string]*Entry
-	mutex   sync.RWMutex
+	log                *zap.Logger
+	buckets            map[string]*Entry
+	mutex              sync.RWMutex
+	PriorityMultiplier uint16
+	Limits             map[LimitType]*Limit
```

Comment:
OK, so the Limiter now has the priority multiplier and a set of limit types. There is the DEFAULT limit and now also a PUBLISH limit. This is set up so that we can add more limit types in the future if we decide to dial down the stream request limits. The `Spend` call now has to pass the limit type to apply. I updated the limit values as we discussed (see above). WDYT?